File: src/tester/runner.ts
import type { TestCase, ModelResult } from "./types.ts";
import { getCases } from "./cases.ts";
import { NetworkSelector } from "./network.ts";
import { CircuitBreaker } from "./circuitBreaker.ts";

export class HTTPStatusError extends Error {
  constructor(public status: number, message: string, public headers?: Headers) {
    super(message);
    this.name = "HTTPStatusError";
  }
}

/**
 * 线程安全的 API 密钥轮换器（带滑动窗口限速）
 */
export class KeyRotator {
  private keys: string[];
  private idx = 0;
  private rateLimit: number; // 每 key 每 window 秒最多请求次数，0 表示不限
  private window: number; // 限速窗口（秒）
  private timestamps = new Map<string, number[]>();

  constructor(keys: string[], rateLimit = 0, window = 60) {
    this.keys = keys.map((k) => k.trim()).filter((k) => k !== "");
    this.rateLimit = rateLimit;
    this.window = window;
    for (const key of this.keys) {
      this.timestamps.set(key, []);
    }
  }

  private availableWait(key: string, now: number): number {
    if (!this.rateLimit) {
      return 0.0;
    }
    const q = this.timestamps.get(key);
    if (!q) {
      return 0.0;
    }
    // 清理窗口外的时间戳
    while (q.length > 0 && now - q[0]! >= this.window) {
      q.shift();
    }
    if (q.length < this.rateLimit) {
      return 0.0;
    }
    // 返回需要等待的秒数
    return this.window - (now - q[0]!) + 0.01;
  }

  private selectKey(): [string, number] {
    const n = this.keys.length;
    if (n === 0) {
      throw new Error("No API keys configured");
    }
    let minWait = Infinity;
    let bestKey = "";
    let bestOffset = 0;
    const now = performance.now() / 1000;

    for (let i = 0; i < n; i++) {
      const key = this.keys[(this.idx + i) % n]!;
      const wait = this.availableWait(key, now);
      if (wait === 0.0) {
        bestKey = key;
        bestOffset = i;
        minWait = 0.0;
        break;
      }
      if (wait < minWait) {
        minWait = wait;
        bestKey = key;
        bestOffset = i;
      }
    }

    // 推进轮询索引（无论是否需要等待都要推进）
    this.idx = (this.idx + bestOffset + 1) % n;

    // 记录时间戳
    if (this.rateLimit && bestKey) {
      const q = this.timestamps.get(bestKey);
      if (q) {
        q.push(now);
      }
    }

    return [bestKey, minWait];
  }

  /**
   * 返回下一个可用的 API Key（在锁外 sleep，避免阻塞其他协程）
   */
  async next(): Promise<string> {
    const [key, wait] = this.selectKey();
    if (wait > 0) {
      console.log(`  [限速] Key ${key.slice(-8)} 等待 ${wait.toFixed(1)}s`);
      await new Promise((resolve) => setTimeout(resolve, wait * 1000));
    }
    return key;
  }
}

/**
 * 简易并发控制信号量
 */
export class Semaphore {
  private activeCount = 0;
  private queue: (() => void)[] = [];

  constructor(private maxConcurrency: number) {}

  async acquire(): Promise<void> {
    if (this.activeCount < this.maxConcurrency) {
      this.activeCount++;
      return;
    }
    return new Promise<void>((resolve) => {
      this.queue.push(resolve);
    });
  }

  release(): void {
    if (this.queue.length > 0) {
      const next = this.queue.shift();
      if (next) {
        next();
      }
    } else {
      this.activeCount--;
    }
  }
}

export class TestRunner {
  private config: any;
  private mode: string;
  private baseUrl: string;
  private chatEp: string;
  private embedEp: string;
  private concurrency: number;
  private retry: number;
  private interval: number;
  private requiredOnly: boolean;
  private progressCallback?: (results: ModelResult[]) => Promise<void> | void;
  private rotator: KeyRotator;
  private selector: NetworkSelector;
  private sem: Semaphore;
  private breaker: CircuitBreaker;

  constructor(config: any, mode: string, progressCallback?: (results: ModelResult[]) => Promise<void> | void) {
    this.config = config;
    this.mode = mode;
    this.baseUrl = (config.api?.base_url || "").replace(/\/+$/, "");
    this.chatEp = config.api?.chat_endpoint || "";
    this.embedEp = config.api?.embeddings_endpoint || "";
    this.concurrency = config.testing?.concurrency || 5;
    this.retry = config.testing?.retry_count || 2;
    this.interval = config.testing?.request_interval || 1;
    this.requiredOnly = !!config.testing?.required_only;
    this.progressCallback = progressCallback;

    const rateLimit = config.testing?.rate_limit_per_key || 0;
    const rateWindow = config.testing?.rate_limit_window || 60;
    this.rotator = new KeyRotator(config.api_keys || [], rateLimit, rateWindow);
    this.selector = new NetworkSelector(config);
    this.sem = new Semaphore(this.concurrency);
    this.breaker = new CircuitBreaker(
      config.testing?.circuit_breaker_threshold || 5,
      config.testing?.circuit_breaker_reset_timeout || 30000,
    );
  }

  private resolveEndpoint(category: string): string {
    const epMap: Record<string, string> = {
      text_embedding: this.embedEp,
      multimodal_embedding: this.embedEp,
      reranker: "/reranking",
      image_generation: "/images/generations",
      audio: "/audio/speech",
    };
    return epMap[category] || this.chatEp;
  }

  private async doRequest(
    url: string,
    payload: any,
    headers: Record<string, string>,
    streaming: boolean,
    clientKwargs: any
  ): Promise<[any, number]> {
    const t0 = performance.now();

    const options: any = {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      ...clientKwargs,
    };

    const resp = await fetch(url, options);
    if (resp.status >= 400) {
      throw new HTTPStatusError(resp.status, `HTTP error! status: ${resp.status}`, resp.headers);
    }

    if (streaming) {
      if (!resp.body) {
        throw new Error("No response body for streaming");
      }
      const reader = resp.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";
      const chunks: string[] = [];
      let ttftMs: number | null = null;

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";

        for (const line of lines) {
          const trimmed = line.trim();
          if (trimmed.startsWith("data: ") && trimmed !== "data: [DONE]") {
            if (ttftMs === null) {
              ttftMs = performance.now() - t0;
            }
            chunks.push(trimmed.slice(6));
          }
        }
      }
      if (buffer) {
        const trimmed = buffer.trim();
        if (trimmed.startsWith("data: ") && trimmed !== "data: [DONE]") {
          if (ttftMs === null) {
            ttftMs = performance.now() - t0;
          }
          chunks.push(trimmed.slice(6));
        }
      }

      const elapsed = performance.now() - t0;
      return [
        {
          success: chunks.length > 0,
          elapsed_ms: Math.round(elapsed),
          ttft_ms: ttftMs ? Math.round(ttftMs) : null,
          chunk_count: chunks.length,
        },
        elapsed,
      ];
    } else {
      const contentType = resp.headers.get("content-type") || "";
      let data: any;
      if (contentType.includes("json")) {
        data = await resp.json();
      } else {
        const arrayBuffer = await resp.arrayBuffer();
        data = new Uint8Array(arrayBuffer);
      }
      const elapsed = performance.now() - t0;
      return [data, elapsed];
    }
  }

  private async runSingleCase(
    model: any,
    category: string,
    caseItem: TestCase,
    clientKwargs: any
  ): Promise<ModelResult> {
    const modelId = model.id || model.model_id || "";
    const ep = this.resolveEndpoint(category);
    const url = `${this.baseUrl}${ep}`;
    const streaming = caseItem.tags.includes("streaming");

    for (let attempt = 0; attempt <= this.retry; attempt++) {
      try {
        if (attempt > 0) {
          await new Promise((resolve) => setTimeout(resolve, this.interval * attempt * 1000));
        }

        // Key 轮换选择（锁外/Semaphore 外）
        const apiKey = await this.rotator.next();

        // 仅在 HTTP 请求执行段申请并发 Semaphore，保持 sleep 独立
        await this.sem.acquire();
        let raw: any;
        let elapsed = 0.0;
        try {
          const [resData, elapsedMs] = await this.breaker.execute(async () => {
            const headers: Record<string, string> = {
              "Authorization": `Bearer ${apiKey}`,
              "Content-Type": "application/json",
              "Accept": streaming ? "text/event-stream" : "application/json",
            };
            const payload = caseItem.buildPayload(modelId);
            return this.doRequest(url, payload, headers, streaming, clientKwargs);
          });
          raw = resData;
          elapsed = elapsedMs;
        } finally {
          this.sem.release();
        }

        const parsed = caseItem.parseResult(raw, elapsed);
        const result: ModelResult = {
          model_id: modelId,
          category,
          case_name: caseItem.name,
          test: caseItem.test,
          status: parsed.success ? "pass" : "fail",
          success: !!parsed.success,
          elapsed_ms: Math.round(elapsed),
          ...parsed,
        };
        return result;
      } catch (e: any) {
        if (e instanceof HTTPStatusError) {
          const status = e.status;
          if (status === 404 || status === 422) {
            return {
              model_id: modelId,
              category,
              case_name: caseItem.name,
              test: caseItem.test,
              status: "skip",
              reason: `HTTP ${status}`,
              success: false,
              elapsed_ms: 0,
            };
          }
          if (status === 429) {
            const retryAfter = e.headers?.get('retry-after');
            const waitMs = retryAfter ? parseInt(retryAfter) * 1000 : (1 + attempt) * 2000;
            console.log(`  [限速] ${modelId} 等待 ${waitMs / 1000}s (attempt ${attempt + 1}/${1 + this.retry})`);
            await new Promise((resolve) => setTimeout(resolve, waitMs));
          } else {
            console.log(`  [${modelId}] ${caseItem.name} HTTP ${status} (attempt ${attempt + 1}/${1 + this.retry})`);
            await new Promise((resolve) => setTimeout(resolve, (1 + attempt) * 1000));
          }
        } else {
          console.log(`  [${modelId}] ${caseItem.name} error: ${e.message || e} (attempt ${attempt + 1}/${1 + this.retry})`);
          await new Promise((resolve) => setTimeout(resolve, 1000));
        }
      }
    }

    return {
      model_id: modelId,
      category,
      case_name: caseItem.name,
      test: caseItem.test,
      status: "error",
      success: false,
      elapsed_ms: 0,
    };
  }

  private async runModel(model: any, category: string, clientKwargs: any): Promise<ModelResult[]> {
    let cases = getCases(category);
    if (this.requiredOnly) {
      cases = cases.filter((c) => c.required);
    }
    const tasks = cases.map((c) => this.runSingleCase(model, category, c, clientKwargs));
    return Promise.all(tasks);
  }

  async runAll(groups: Record<string, any[]>): Promise<Record<string, ModelResult[]>> {
    const clientKwargs = this.selector.buildClientKwargs(this.mode);
    const allResults: Record<string, ModelResult[]> = {};
    const totalModels = Object.values(groups).reduce((sum, list) => sum + list.length, 0);
    let done = 0;

    for (const [category, models] of Object.entries(groups)) {
      console.log(`  🔬 [${category}] 开始测试 ${models.length} 个模型...`);
      const catResults: ModelResult[] = [];
      const modelPromises = models.map(async (m, idx) => {
        const modelResults = await this.runModel(m, category, clientKwargs);
        catResults[idx] = modelResults;
        done++;
        const modelId = m.id || m.model_id || "?";
        const passed = modelResults.filter((r) => r.status === "pass").length;
        const totalC = modelResults.length;
        console.log(`    [${done}/${totalModels}] ${modelId}: ${passed}/${totalC} 用例通过`);
        if (this.progressCallback) {
          await this.progressCallback(modelResults);
        }
      });
      await Promise.all(modelPromises);
      allResults[category] = catResults.flat();
    }

    return allResults;
  }
}

File: python_legacy/tester/runner.py
# -*- coding: utf-8 -*-
"""
测试执行引擎
- 并发控制（semaphore）
- 多密钥轮换
- 流式响应特殊处理
- 失败重试
"""

import asyncio
import logging
import time
from typing import Any, Dict, List

import httpx

from tester.cases import TestCase, get_cases
from tester.network import NetworkSelector

logger = logging.getLogger(__name__)


import collections


class KeyRotator:
    """线程安全的 API 密钥轮换器（带滑动窗口限速）"""

    def __init__(self, keys: List[str], rate_limit: int = 0, window: int = 60):
        self._keys = [k.strip() for k in keys if k.strip()]
        self._idx = 0
        self._lock = asyncio.Lock()
        self._rate_limit = rate_limit        # 每 key 每 window 秒最多请求次数，0 表示不限
        self._window = window                # 限速窗口（秒）
        self._timestamps: Dict[str, collections.deque] = {
            k: collections.deque() for k in self._keys
        }

    def _available_wait(self, key: str) -> float:
        """返回 key 当前需要等待的秒数（0 表示可立即使用）"""
        if not self._rate_limit:
            return 0.0
        q = self._timestamps[key]
        now = time.monotonic()
        while q and now - q[0] >= self._window:
            q.popleft()
        if len(q) < self._rate_limit:
            return 0.0
        return self._window - (now - q[0]) + 0.01

    async def _select_key(self) -> tuple:
        """
        在锁内选择最优 key，返回 (key, wait_secs)。
        不在锁内执行 sleep，避免阻塞其他协程。
        """
        async with self._lock:
            n = len(self._keys)
            min_wait = float("inf")
            best_key = None
            best_offset = 0

            for i in range(n):
                key = self._keys[(self._idx + i) % n]
                wait = self._available_wait(key)
                if wait == 0.0:
                    best_key = key
                    best_offset = i
                    min_wait = 0.0
                    break
                if wait < min_wait:
                    min_wait = wait
                    best_key = key
                    best_offset = i

            # 推进轮询索引（无论是否需要等待都要推进，修复 BUG-5）
            self._idx = (self._idx + best_offset + 1) % n

            # 记录时间戳（在 sleep 前记录，避免多协程同时获得同一 key）
            if self._rate_limit and best_key:
                self._timestamps[best_key].append(time.monotonic())

            return best_key, min_wait

    async def next(self) -> str:
        """返回下一个可用的 API Key（在锁外 sleep，避免阻塞其他协程）"""
        key, wait = await self._select_key()
        if wait > 0:
            logger.debug(f"  [限速] Key {key[-8:]!r} 等待 {wait:.1f}s")
            await asyncio.sleep(wait)
        return key


class TestRunner:
    """并发执行所有分组的测试用例"""

    def __init__(self, config: dict, mode: str, progress_callback=None):
        self.config = config
        self.mode = mode
        self.base_url = config["api"]["base_url"].rstrip("/")
        self.chat_ep = config["api"]["chat_endpoint"]
        self.embed_ep = config["api"]["embeddings_endpoint"]
        self.concurrency = config["testing"]["concurrency"]
        self.retry = config["testing"]["retry_count"]
        self.interval = config["testing"]["request_interval"]
        self.required_only = config["testing"].get("required_only", False)
        self.progress_callback = progress_callback
        rate_limit = config["testing"].get("rate_limit_per_key", 0)
        rate_window = config["testing"].get("rate_limit_window", 60)
        self.rotator = KeyRotator(config["api_keys"], rate_limit, rate_window)
        self._selector = NetworkSelector(config)
        # D-6: Semaphore 延迟创建，避免 Python 3.12 跨事件循环警告
        self._sem: asyncio.Semaphore | None = None

    def _resolve_endpoint(self, category: str) -> str:
        """根据模型分类选择请求端点"""
        ep_map = {
            "text_embedding": self.embed_ep,
            "multimodal_embedding": self.embed_ep,
            "reranker": "/reranking",
            "image_generation": "/images/generations",
            "audio": "/audio/speech",
        }
        return ep_map.get(category, self.chat_ep)

    async def _do_request(
        self,
        client: httpx.AsyncClient,
        url: str,
        payload: dict,
        headers: dict,
        streaming: bool = False,
    ) -> tuple[dict | bytes, float]:
        """执行单次 HTTP 请求，返回 (response_data, elapsed_ms)"""
        t0 = time.perf_counter()

        if streaming:
            # 流式模式：收集 TTFT 和完整内容
            chunks = []
            ttft_ms = None
            async with client.stream("POST", url, json=payload, headers=headers) as resp:
                if resp.status_code >= 400:
                    # 抛出异常以触发外层重试逻辑
                    resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if line.startswith("data: ") and line != "data: [DONE]":
                        if ttft_ms is None:
                            ttft_ms = (time.perf_counter() - t0) * 1000
                        chunks.append(line[6:])
            elapsed = (time.perf_counter() - t0) * 1000
            return {
                "success": len(chunks) > 0,
                "elapsed_ms": round(elapsed, 0),
                "ttft_ms": round(ttft_ms, 0) if ttft_ms else None,
                "chunk_count": len(chunks),
            }, elapsed

        else:
            resp = await client.post(url, json=payload, headers=headers)
            elapsed = (time.perf_counter() - t0) * 1000
            # 触发 HTTPStatusError，让外层重试逻辑捕获 403/429/500 等错误
            resp.raise_for_status()
            content_type = resp.headers.get("content-type", "")
            if "json" in content_type:
                return resp.json(), elapsed
            elif resp.content:
                return resp.content, elapsed
            return {}, elapsed


    async def _run_single_case(
        self,
        model: Dict[str, Any],
        category: str,
        case: TestCase,
        client: httpx.AsyncClient,
    ) -> Dict[str, Any]:
        """执行单个测试用例，包含重试逻辑"""
        model_id = model.get("id", "")
        ep = self._resolve_endpoint(category)
        url = f"{self.base_url}{ep}"
        streaming = "streaming" in case.tags

        for attempt in range(1 + self.retry):
            try:
                await asyncio.sleep(self.interval * attempt)
                api_key = await self.rotator.next()
                async with self._sem:
                    headers = {
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json",
                        "Accept": "text/event-stream" if streaming else "application/json",
                    }
                    payload = case.build_payload(model_id)

                    raw, elapsed = await self._do_request(
                        client, url, payload, headers, streaming
                    )

                result = case.parse_result(raw, elapsed)
                result["model_id"] = model_id
                result["category"] = category
                result["case_name"] = case.name
                result["attempt"] = attempt + 1
                result["status"] = "pass" if result.get("success") else "fail"
                return result

            except httpx.HTTPStatusError as e:
                status = e.response.status_code
                # 404 模型不存在 / 422 请求格式错误 → 无意义重试
                if status in (404, 422):
                    return {
                        "model_id": model_id,
                        "category": category,
                        "case_name": case.name,
                        "test": case.test,
                        "status": "skip",
                        "reason": f"HTTP {status}",
                        "success": False,
                    }
                # 403 表示当前 Key 无权限，换下一个 Key 重试
                # 429 限流 / 500 服务器错误 → 同样重试
                logger.debug(f"  [{model_id}] {case.name} HTTP {status} (attempt {attempt+1}/{1+self.retry})")
                await asyncio.sleep(1 + attempt)   # 退避性等待
            except Exception as e:
                logger.debug(f"  [{model_id}] {case.name} error: {e} (attempt {attempt+1})")
                await asyncio.sleep(1)

        return {
            "model_id": model_id,
            "category": category,
            "case_name": case.name,
            "test": case.test,          # 修复：error 路径同样需要 test 字段参与评分
            "status": "error",
            "success": False,
        }

    async def _run_model(
        self, model: Dict[str, Any], category: str,
        client: httpx.AsyncClient,
    ) -> List[Dict[str, Any]]:
        """对单个模型运行所属分类的全部测试用例"""
        cases = get_cases(category)
        if self.required_only:
            cases = [c for c in cases if c.required]
        tasks = [self._run_single_case(model, category, c, client) for c in cases]
        return await asyncio.gather(*tasks)

    async def run_all(
        self, groups: Dict[str, List[Dict[str, Any]]]
    ) -> Dict[str, Any]:
        """运行所有分组的全部模型测试（复用单一 httpx 客户端）"""
        # 延迟创建 Semaphore，确保在当前事件循环内创建（修复 D-6）
        if self._sem is None:
            self._sem = asyncio.Semaphore(self.concurrency)
        client_kwargs = self._selector.build_client_kwargs(self.mode)
        all_results: Dict[str, Any] = {}
        total_models = sum(len(v) for v in groups.values())
        done = 0

        async with httpx.AsyncClient(**client_kwargs) as client:  # 复用连接池
            for category, models in groups.items():
                logger.info(
                    f"  🔬 [{category}] 开始测试 {len(models)} 个模型..."
                )
                cat_results = []
                tasks = [self._run_model(m, category, client) for m in models]
                for coro in asyncio.as_completed(tasks):
                    model_results = await coro
                    cat_results.extend(model_results)
                    done += 1
                    model_id = model_results[0]["model_id"] if model_results else "?"
                    passed = sum(1 for r in model_results if r["status"] == "pass")
                    total_c = len(model_results)
                    logger.info(
                        f"    [{done}/{total_models}] {model_id}: "
                        f"{passed}/{total_c} 用例通过"
                    )
                    if self.progress_callback:
                        await self.progress_callback(model_results)
                all_results[category] = cat_results

        return all_results
